test/test_sklearn_compat.py

```diff
@@ -68,5 +81,343 @@ def test_mmc(self):
+  (X, tuples, y, tuples_train, tuples_test,
+   y_train, y_test, preprocessor) = build_dataset(preprocessor)
+  estimator = clone(estimator)
+  estimator.set_params(preprocessor=preprocessor)
+  if hasattr(estimator, "num_dims"):
+    estimator.num_dims = 1
+  estimator.fit(tuples, y)
+  for method in ["predict", "decision_function", "predict_proba"]:
+    if hasattr(estimator, method):
+      dict_before = estimator.__dict__.copy()
+      getattr(estimator, method)(tuples)
+      assert estimator.__dict__ == dict_before, \
+          ("Estimator changes __dict__ during %s"
+           % method)
+  for method in ["transform"]:
+    if hasattr(estimator, method):
+      dict_before = estimator.__dict__.copy()
+      # we transform only 2D arrays (dataset of points)
+      getattr(estimator, method)(X)
+      assert estimator.__dict__ == dict_before, \
+          ("Estimator changes __dict__ during %s"
+           % method)
+
+
+@pytest.mark.parametrize('preprocessor', [None, build_data()[0]])
+@pytest.mark.parametrize('estimator, build_dataset', list_estimators,
+                         ids=ids_estimators)
+def test_dont_overwrite_parameters(estimator, build_dataset, preprocessor):
+  # Adapted from scikit-learn
+  # check that fit method only changes or sets private attributes
+  (X, tuples, y, tuples_train, tuples_test,
+   y_train, y_test, preprocessor) = build_dataset(preprocessor)
+  estimator = clone(estimator)
+  estimator.set_params(preprocessor=preprocessor)
+  if hasattr(estimator, "num_dims"):
+    estimator.num_dims = 1
+  dict_before_fit = estimator.__dict__.copy()
+
+  estimator.fit(tuples, y)
+  dict_after_fit = estimator.__dict__
+
+  public_keys_after_fit = [key for key in dict_after_fit.keys()
+                           if is_public_parameter(key)]
+
+  attrs_added_by_fit = [key for key in public_keys_after_fit
+                        if key not in dict_before_fit.keys()]
+
+  # check that fit doesn't add any public attribute
+  assert not attrs_added_by_fit, (
+      "Estimator adds public attribute(s) during"
+      " the fit method."
+      " Estimators are only allowed to add private "
+      "attributes"
+      " either started with _ or ended"
+      " with _ but %s added" % ', '.join(attrs_added_by_fit))
+
+  # check that fit doesn't change any public attribute
+  attrs_changed_by_fit = [key for key in public_keys_after_fit
+                          if (dict_before_fit[key]
+                              is not dict_after_fit[key])]
+
+  assert not attrs_changed_by_fit, (
+      "Estimator changes public attribute(s) during"
+      " the fit method. Estimators are only allowed"
+      " to change attributes started"
+      " or ended with _, but"
+      " %s changed" % ', '.join(attrs_changed_by_fit))
+
+
+def _get_args(function, varargs=False):
+    """Helper to get function arguments"""
+
+    try:
+        params = signature(function).parameters
+    except ValueError:
+        # Error on builtin C function
```

Comment:
Could you clarify this a bit more?

Reply:
Actually I copied this function from scikit-learn and forgot to add in the comment "Adapted from scikit learn"... But I don't know why I copied it, I'll import it instead.